superduperdb/components/datatype.py

```diff
@@ -692,47 +669,27 @@ class File(_BaseEncodable):
 
     :param x: path to the file
     :param file_name: File name
+    :param file: file reference, format `&:file:{file_name}/{file_id}
     """
 
     leaf_type: t.ClassVar[str] = 'file'
     x: t.Any = Empty()
     lazy: t.ClassVar[bool] = False
     file_name: t.Optional[str] = None
+    file: dc.InitVar[t.Optional[bytearray]] = None
 
-    def __post_init__(self, db):
+    def __post_init__(self, db, file=None):
         super().__post_init__(db)
+        self._file = file
 
         if not self.lazy and isinstance(self.x, Empty):
             self.init()
 
-    @classmethod
-    def build_from_reference(cls, reference: str, datatype: DataType):
-        """Build an instance of the File from a reference.
-
-        :param reference: The reference to build the instance from.
-                        e.g. `:file:file/123/hello.txt`
-        :param datatype: The datatype of the instance.
-        """
-        *_, file_id, file_name = reference.split('/')
-        file_name = file_name.replace('__', '.')
-        file = cls(
-            file_id=file_id,
-            file_name=file_name,
-            datatype=datatype,
-        )
-        if not file.lazy:
-            return file.x
-        return file
-
     @property
-    def id(self):
-        """Get the id of the object.
-
-        We can use this id as a reference to rebuild the object.
-        """
+    def _id(self):
         assert self.file_id is not None
         file_name = self.file_name.replace('.', '__')
-        return f':file:{self.leaf_type}/{self.file_id}/{file_name}'
+        return f':file:{file_name}/{self.file_id}'
```

Comment:
Do we have a general schema for creating `self._id` for `Leaf` instances?

Reply:
There should be one. It is possible to add a class variable such as ‘add_reference_tag’.

This is because we need to handle cases where the leaf types like artifact/lazy_artifact are different, type_ids are different, but the reference tags are the same.

However, I prefer the return to be the same as self._id, and to support overriding this method, so that each class can construct this string according to its own situation and can parse this string to construct itself.